### ISSUE TITLE
Healthcheck endpoint

### ISSUE THREAD

**Issue by pando85** [kanidm/kanidm#216]
We could have a simple endpoint that check kanindmd status and return 200. I like `/v1/ping` which return 200 with a body of `pong` but anything could be useful.

**Comment by Firstyear** [comment]
https://openmetrics.io/ was suggested by @pando85 as a standard we could consider following, so we should look at this for status.

**Comment by yaleman** [comment]
For people looking for this in the future, there's a simple health check at `/status` added in #221

**Comment by Firstyear** [comment]
I think this issue now is more about extending the endpoint to internally verify more parts.

**Comment by matthew-salerno** [comment]
Just gonna through my opinion here since this is still in dev. I just read over openmetrics. I like it quite a bit. Not a fan at how hard it is to read, but it makes sense under their design constraints for high throughput of metrics. That being said, idk how well it lines up with our use case here. I imagine for detailed info people would use the logs and maybe an aggregator. We also probably don't want to expose too much over http. So all I can think of is an alive status, the configured uri, maybe a version number? At that low of a data volume we might as well use json, yaml, or ron, and then at least it will be easy to curl.

**Comment by yaleman** [comment]
If you're monitoring the metrics of an IdP, then ingesting prometheus metrics (which is the base implementation of OpenMetrics) should be trivial - it *can* be a curl request and it gives you a list of key-value pairs of `metric value`, ie `memoryUsed 123456778`.

In the current alpha stage there's less to monitor, but as soon as any replication's occurring then there's plenty of things to capture for monitoring purposes that would be handy to expose for specific monitoring outside logs.

Don't get me wrong, there's plenty of ways to achieve monitoring through logs - that's literally my day job 😇 - but from an ops perspective, systems collating and exposing their metrics is super handy.

**Comment by Firstyear** [comment]
well, there is also stuff related to cache behaviour and efficiency too that we could be capturing, but I just haven't exposed all of that either ... 

But yes, thanks for your comments @matthew-salerno, I think early on I'd probably just make a json end point, but as @yaleman points out, once we have more systems in a large topology it becomes a bit more interesting. 

We may even wanth to start to track stats of how many unique operations are occuring as well.

**Comment by yaleman** [comment]
> A slightly different _http_ endpoint may be nice for Kubernetes probes (https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/). I don't think it would work with https, especially with a self-signed or private root CA signed cert. It's probably possible to use command probes today, but that feels less good than an endpoint.

If you configure a HTTPS probe, [the default is to skip verification](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/), so it's not an issue.

> If scheme field is set to HTTPS, the kubelet sends an HTTPS request skipping the certificate verification.

**Comment by jfroy** [comment]
🥲 I read that page several times and I somehow managed to glaze over that.

**Comment by yaleman** [comment]
No worries, easy to do, I originally went looking for the options to set to skip it 😄

**Comment by tumbl3w33d** [comment]
Not sure this is the right ticket for it, but I'm curious if there has been further development on this end. I've used the [tracing](https://docs.rs/crate/tracing/latest) lib(s) in a rust application myself and was quite happy with it. It optionally comes with an [opentelemetry](https://docs.rs/tracing-opentelemetry/0.19.0/tracing_opentelemetry/) integration. After all, this gives you all kinds of detailed log magic, like easily switching to json logs etc., plus metrics in a standard format.

I'd be interested in collecting metrics and logs from whatever endpoint provided using elastic agent. It has an integration for [opentelemetry](https://www.elastic.co/guide/en/apm/guide/current/open-telemetry.html) but can basically, with [more](https://www.elastic.co/guide/en/integrations-developer/current/what-is-an-integration.html) or less (there are a few common "custom" integrations) effort, pick up anything with both notions - opening a socket where the app can send data or poll something like a prometheus endpoint.

Not trying to push this into some vendor-specific direction, just adding some thoughts and a perspective of how I would use it.

**Comment by yaleman** [comment]
I have plans to investigate Otel metrics/traces, but they're not in place yet.

**Comment by Firstyear** [comment]
We need to investigate otel to see if it's going to work properly in the async context of kani since we have some strict requirements on how we deal with thread safety. So it's pending my review there.

**Comment by Firstyear** [comment]
Relates #2183, this should reflect the data in the fault manager.

**Comment by tumbl3w33d** [comment]
I noticed today that it would be great to have a monitoring endpoint that also gives information about replication health. @yaleman mentioned that in a way with

> but as soon as any replication's occurring then there's plenty of things to capture for monitoring purposes that would be handy to expose for specific monitoring outside logs

It could be either part of a json payload returned from `/status` or a separate endpoint. It is true that I could detect that from watching logs and creating alerts built on top of that, but monitoring systems like nagios/icinga work better with such endpoint that returns the information. My replication recently broke unnoticed because the certificates expired.

On a side note: The documentation currently states

> kanidmd currently responds to HTTP GET requests at the /status endpoint with a JSON object of either "true" or "false"

which is not the case. The response is `text/plain` with content `true` or `false`.

**Comment by yaleman** [comment]
The replication status thing is definitely on the radar, probably around the time that I get the metrics internals working because that'll make it easier to store/parse/serve.

**Comment by yaleman** [comment]
I’m not a huge fan of this idea; TBH any access to the API should be logged, and it’s a relatively tiny amount of logging. (It should be slightly more terse, but that something we need to work on in general)

**Comment by Firstyear** [comment]
We've minimised the status log output to be as little as possible - a single line. I think that's about as good as we can make it.

**Comment by felix** [comment]
No worries. The docs list info, debug, and trace as the only levels. I went looking for a way to _reduce_ the log output for some clarity.

**Comment by Firstyear** [comment]
@felix we can't reduce it any more is the problem - if we do, we break a ton of stuff. It's to do with how our logging spans work.

**Comment by Papierkorb** [comment]
Hello, chiming in as the kanidm docs linked here.

I've noticed that the `/status` endpoint only relies on returning true or false. However, is my reading of the code correct that it also returns `200 OK` even if it responds with "false"?

This would make it much harder to integrate into simpler health checkers like that one of Kubernetes, which does a HTTP GET request and then expects the HTTP response code to reflect good or bad health: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/

Would it be possible to add this? Sorry if the code already does it, I'm not familiar with axum.

**Comment by Firstyear** [comment]
So here is the issue - do we use the response code as the signal of health? Or is the response code the signal of "your request to get the status worked". 

On one side, if we make the response code 500 or something, what does this mean? We failed to generate a status page? Or that the service has a fault?

If we set it to 200 we are saying we successfully gave you a status page, even though the status may be that the service is unhealthy? Or are we saying that the service as a whole is healthy? Then what's the point of the body of the page? 

That's the problem with HTTP response codes, they can signal two different things - that the *request itself* succeeded, even if the body of the request has a meaning otherwise, or that people rely on them to signal the nature of the body. Within Kanidm we have chosen that the response codes indicate the success of the request, and the body requires interpretation to it's meaning. 

In this case, response 200 is correct in my opinion - even if we were in an unhealthy state, we are correctly returning a status page, even if that status page body is indicating service faults.

**Comment by Kernald** [comment]
While I agree that a 200 _feels_ correct, it goes against what most services (from my anecdotal experience) seem to expose. Probe built into tools like Consul tend to not even let you compare the body, they only let you check the return code. While it's pretty trivial to implement a custom, curl-based healtcheck for Kanidm, it points to the _expectation_ of the return code to match the service status, rather than the request handling status.

**Comment by Firstyear** [comment]
In this case, if there is a server issue, then I think the right status with that status information should be 503 potentially for service unavailable. But I think the json structure of that response body should be similar regardless of 200/503.

**Comment by Papierkorb** [comment]
@kernald that's what I have in mind. I run kanidm in kubernetes which supports a simple http request to check the status code for healthiness. Many devices seem to expose a `/health` similar to kanidm with a lot of metrics, and a `/healthz` which just returns a status code to indicate a binary "all good" or "something is not right". That's exactly what you want, as when the service is inoperable you don't care if the http service is still responding. If the http service is responding to http then obviously it's fine, and if it's not then, again, the service is inoperable and needs to be looked at (if not automatically restarted).

@Firstyear agree, the response should include a (simple) JSON response in addition. It could contain a simple boolean or more details.

**Comment by Jak2k** [comment]
Would it be possible to add Prometheus support?

**Comment by pando85** [comment]
Another important point that hasn’t been considered yet is the readiness probe.
Returning a 200 status is fine for indicating that the server is up, but it doesn’t necessarily mean the instance is ready to handle requests.

I've identified two scenarios where the instance should **not** be considered ready:

- When the `admin/idm_admin` credentials have not been set. In this case, the instance has no data initialized, so it shouldn’t be part of the load balancer pool.
- When replication is failing and requires manual intervention. Here, the instance’s data may be outdated, and adding it to the pool could lead to inconsistent or unexpected behavior.

**Comment by yaleman** [comment]
The replication ones sound like good checks (if replication is enabled), will have to work out the combinations of states for checking the admin creds

**Comment by Firstyear** [comment]
Yes I planned to add these soon as part of the ongoing work I'm doing.

**Comment by pando85** [comment]
If you want some help here, I can contribute with some guidelines.

**Comment by indibart** [comment]
Could a `kanidmd healthcheck` subcommand be added that hits `/status` over loopback and exits `0`/`1`? 

When using Docker/Podman `HEALTHCHECK` you need to execute a command *in* the container. But as the `kanidmd` image is deliberately minimal there also is no shell, no `curl`/`wget`, not even `busybox`. So there's no way to call `/status` from inside the container itself.

**Comment by yaleman** [comment]
> Could a `kanidmd healthcheck` subcommand be added that hits `/status` over loopback and exits `0`/`1`? 
> 
> When using Docker/Podman `HEALTHCHECK` you need to execute a command *in* the container. But as the `kanidmd` image is deliberately minimal there also is no shell, no `curl`/`wget`, not even `busybox`. So there's no way to call `/status` from inside the container itself.

It's there under scripting healthcheck and should already be part of the container defs

**Comment by Firstyear** [comment]
Yes, it's already part of the container, and we have a /status endpoint - it could certainly do more, but it's there.
